### PR TITLE
Enhance errors handling in issuer-api

### PR DESCRIPTION
**Pull request by SuperBatata**
## Description

Provide a clear and concise description of the changes made in this pull request.

Fixes #[issue number]

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 1]
```diff
@@ -221,26 +237,61 @@ fun Application.issuerApi() {
...
+                                description = "Bad request"
+                            }
                         }
                     }) {
-
                         val body = context.receive<Map<String, JsonElement>>()
+                        val keyJson = body["issuerKey"]?.jsonObject
+                            ?: return@post context.respond(
+                                HttpStatusCode.BadRequest,
+                                "Missing issuerKey in the request body."
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:249]: Look how this message is much more specific than the one at line 121 ("Issuer key must not be empty")

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 2]
```diff
@@ -221,26 +237,61 @@ fun Application.issuerApi() {
                                     )
                                 }
                             }
+                            "400" to {
+                                description = "Bad request"
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:241]: That's all we can provide? Any more detail?

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 3]
```diff
@@ -221,26 +237,61 @@ fun Application.issuerApi() {
...
+                            )
 
-                        // Sign VC
-                        val jws = vc.signJws(
-                            issuerKey = key, issuerDid = issuerDid, subjectDid = subjectDid
+                        val credentialData = body["credentialData"]?.let {
+                            Json.encodeToString(it)
+                        } ?: return@post context.respond(
+                            HttpStatusCode.BadRequest,
+                            "Missing credentialData in the request body."
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:274]: That's what I was talking about in the comment I left for line 111.

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 4]
```diff
@@ -285,14 +336,28 @@ fun Application.issuerApi() {
                                     }
                                 }
                             }
+                            "400" to {
+                                description =
+                                    "Bad request - The request could not be understood or was missing required parameters."
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:341]: Will it be complemented later in the code flow with more informations?

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 5]
```diff
@@ -449,3 +527,23 @@ fun Application.issuerApi() {
...
 }
+
+private suspend fun PipelineContext<Unit, ApplicationCall>.throwError(
+    it: Throwable
+) {
+    when (it) {
+        is BadRequestException -> context.respond(HttpStatusCode.BadRequest, it.message ?: "Bad request.")
+        is JedisConnectionException -> context.respond(
+            HttpStatusCode.InternalServerError,
+            "Distributed session management couldn't be initialized : Cannot connect to redis server."
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:538]: What if we include the address it tried to use to connect?

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 6]
```diff
@@ -449,3 +527,23 @@ fun Application.issuerApi() {
...
+        )
+
+        is JedisAccessControlException -> context.respond(
+            HttpStatusCode.InternalServerError,
+            "Distributed session management couldn't be initialized : Cannot access redis server, wrong username/password."
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:543]: I imagine a scenario in which the user sets up the access credentials in the wrong place. For them, the credentials are correct but, as the system is looking elsewhere, they may not understand why the error occurred.

So, my first reaction would be to suggest including the username and password that was used here in the response, but I don't think that's a good idea. Any ideas on how to make it clear to the user which user/password is wrong in such a way he can quickly identify the source of the problem?

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt [hunk 7]
```diff
@@ -449,3 +527,23 @@ fun Application.issuerApi() {
...
+        )
+
+        else -> context.respond(HttpStatusCode.InternalServerError, it.message ?: "Failed to issue credentials.")
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:546]: Nothing else that could be included?

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/OnboardingRequest.kt [hunk 8]
```diff
@@ -22,3 +24,15 @@ data class OnboardingRequest(
         method = "jwk"
     )
 )
+
+fun validateOnboardingRequest(request: OnboardingRequest) {
+    require(request.key.backend.isNotEmpty()) {
+        throw WebException(HttpStatusCode.BadRequest, "Key backend must not be empty")
```
Comment (alegomes) [inline waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/OnboardingRequest.kt:30]: I would go for the property name. "Missing property X" or "Property x in the request body cannot be empty"

File: waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt (hunk 6 above)
Comment (SuperBatata) [reply waltid-services/waltid-issuer-api/src/main/kotlin/id/walt/issuer/issuance/IssuerApi.kt:543]: i think its already clear with : Cannot access redis server, wrong username/password

**Review by mikeplotean — APPROVED** [review]

**Pull request merged**